qiskit_experiments/curve_analysis/grouped_curve_analysis.py

```diff
@@ -0,0 +1,335 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Analysis class for multi-group curve fitting.
+"""
+# pylint: disable=invalid-name
+
+from typing import Dict, List, Tuple, Optional, Any
+from collections import defaultdict
+
+import lmfit
+import numpy as np
+from uncertainties import unumpy as unp, UFloat
+
+from qiskit_experiments.framework import ExperimentData, AnalysisResultData, Options
+from qiskit_experiments.exceptions import AnalysisError
+
+from .base_curve_analysis import BaseCurveAnalysis, PARAMS_ENTRY_PREFIX
+from .curve_data import CurveFitResult
+from .utils import analysis_result_to_repr, eval_with_uncertainties
+
+
+class MultiGroupCurveAnalysis(BaseCurveAnalysis):
```

Comment:
Another thing I wonder about is if it should be easy to use a `CurveAnalysis` class in a `MultiGroupCurveAnalysis` class. Other than specifying the groups, the only difference seems to be the optional `_create_composite_analysis_results` method. For example, if one wants to run three separate sinusoidal measurements together, could there be a way to use OscillationAnalysis and name the group keys rather than respecify the model and the fit guess logic from OscillationAnalysis?  I am just wondering. I don't think it is critical if there is not an easy way. With all the recent fitting work, making a new class is not that much work (and some of the initial guess logic is factored out into reusable functions already, making it even easier to specify the models and initial guess method).

Reply:
Thanks, indeed this is what I was thinking of at the beginning, i.e

```python
analysis = MultiGroupGurveAnalysis(
    analyses=[OscillationAnalysis(name="curve1"), OscillationAnalysis(name="curve2")],
    group_keys = {"curve1": {"option": 1}, "curve2": {"option": 2}},
)
```
Seems like this is much cleaner, but actually implementation of drawer is challenging because the logic to draw figure is in each `CurveAnalysis` (so this generates two separate figure), but `MultiGroupGurveAnalysis` may want a single figure with two lines in the same canvas. This looked bit complicated and I gave up to simplify the logic, but I'll look into this direction if I can come up with some clean logic.